### PR TITLE
Clear variables set by CI builds

### PR DESCRIPTION
**Pull request by sharwell**
Fixes [AB#1220549](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1220549)

### CODE REVIEW

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (jaredpar) [inline src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:361]: Comment is saying what, I would focus on "why". Essentially the goal of arcade is to isolate our unit test experience to the exact SDK that we specify in global.json. In the case of VS integration tests though that is the exact opposite of what we want. We want VS to use the SDK that comes with VS, not what arcade thinks is the correct SDK.

**Review by jaredpar — APPROVED** [review]

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:361]: Updated. This what you were after?

**Comment by sharwell** [comment]
Marking as draft since this needs to be rebased to only include the last commit before merge

**Review by jasonmalinowski — APPROVED** [review]

**Comment by sharwell** [comment]
@JoeRobich Is there any way you can figure out what error is in the activity log?

image

**Comment by sharwell** [comment]
@jasonmalinowski It seems clear that NuGet restore is simply not working on CI:

image

**Comment by jasonmalinowski** [comment]
@sharwell Odd. I don't think anything changed on our side there. Do we think this PR is working at this point or do we need to skip tests? We're well beyond our deadline of getting these fixed by now or disabling.

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (jasonmalinowski) [reply src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:362]: Then they wouldn't apply if you're running from the VS test runner. IMHO the script is almost never the place to put things.

**Review by jasonmalinowski — APPROVED** [review]

**Comment by sharwell** [comment]
> We're well beyond our deadline of getting these fixed by now or disabling.

Agreed, except when I tried to disable the tests it still failed. I can disable more and see if it passes.

**Comment by sharwell** [comment]
@jasonmalinowski I updated the test failure handler to capture activity logs, and capture event logs for the 64-bit OOP process. It should be informative even if it's not the last change.

**Comment by sharwell** [comment]
```
System.InvalidOperationException: Connection could not be established with the client. Client did not return a valid connnection. Client: C#/Visual Basic Language Server Client
   at Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.<InitializeAsync>d__81.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.<ActivateLanguageClientAsync>d__87.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)</description>
```

**Comment by sharwell** [comment]
@jasonmalinowski I fixed the issue with tests failing saying the light bulb wasn't present.

@AmadeusW it looks like https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform/pullrequest/268277 caused a pretty severe regression

**Comment by sharwell** [comment]
@jasonmalinowski and now I believe I have also fixed the issue of SDK resolution failing

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (sharwell) [inline src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:630]: @AmadeusW The suggested actions source is getting disposed without first cancelling in-flight calls to `GetSuggestedActionCategoriesAsync`. This seems like a bug.

**Comment by sharwell** [comment]
Finally found the remaining case and will have it fixed shortly

image

File: src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs
Comment (jasonmalinowski) [reply src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:630]: @sharwell Can we file and link to a bug here?

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (jasonmalinowski) [inline src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:369]: If we're running the CI script locally is this going to be set? Wasn't sure if some heuristic like "There's a .dotnet directory and the parent contains Roslyn.sln", but I can imagine that heuristic going bad in other ways.

**Review by jasonmalinowski — APPROVED** [review]
No comments here are critical and can be dealt with as a mop-up later.

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:85]: 1. Length might be stuck. 6 characters is what we can do without hitting max path, at least in the past.

File: src/VisualStudio/IntegrationTest/TestUtilities/ActivityLogCollector.cs
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/ActivityLogCollector.cs:20]: yes, but it's still usable

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (sharwell) [reply src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:369]: This isn't set locally, but neither are the bad environment variables.

**Comment by sharwell** [comment]
@RikkiGibson I figured out the primary source of bugs in `ChangedText.Merge`. Minimal fix here to unblock CI.

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:385]: 📝 This was failing to handle the case where the first remaining old change was an adjusted change.

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:341]: 📝 This case was failing to handle multiple overlaps, e.g.:

```
Old: [--------]
New: [---]  [------]
```

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:352]: 📝 I'm not sure if this case was originally correct or not, but I'll review it in more detail now that CI is running.

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:348]: 📝 Most of the complexity of this change involves the new ability to `goto nextNewChange` without overwriting the value of `oldChange`, since the value might be a subset of the change in `oldChanges[oldIndex]`.

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:337]: 📝 I don't think this delta calculation was correct

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:335]: 📝 This was equivalent to just:

```
var oldChangeReduction = oldChange.NewLength;
```

File: src/Compilers/Core/Portable/Text/ChangedText.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/Text/ChangedText.cs:337]: 📝 Due to the `if` statement conditions, this was equivalent to:

```csharp
var oldChangeReduction = oldChange.NewLength;
AddRange(list, new TextChangeRange(oldChange.Span, 0));
oldDelta = oldDelta - oldChange.Span.Length;
```

📝 I don't think this delta calculation was correct

File: src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs
Comment (jasonmalinowski) [reply src/VisualStudio/IntegrationTest/TestUtilities/VisualStudioInstanceFactory.cs:85]: If point 1 is the reason we're not taking this advice, please add a comment to that effect. (Later. :smile:)

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler @CyrusNajmabadi I need two reviews for the compiler change to unblock our CI. I'm assuming my changes will eventually be replaced by @RikkiGibson's change but for now we just need something that passes.

**Comment by sharwell** [comment]
@jaredpar I'm happy to revert the compiler change if y'all don't want it. Merged to unblock the team.

**Pull request merged**
